File: Design Patterns/Structural/Proxy/proxy_test.go
package proxy

import "testing"

func TestApplicationHandleRequest(t *testing.T) {
	app := &Application{}
	result, err := app.HandleRequest("http://example.com")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if result != "http://example.com - Success" {
		t.Errorf("expected 'http://example.com - Success', got '%s'", result)
	}
}

func TestProxyAllowsRequestsWithinLimit(t *testing.T) {
	app := &Application{}
	p := NewProxy(app, 3)

	for i := 1; i <= 3; i++ {
		result, err := p.HandleRequest("http://example.com")
		if err != nil {
			t.Fatalf("request %d should succeed, got error: %v", i, err)
		}
		if result != "http://example.com - Success" {
			t.Errorf("request %d: expected success response, got '%s'", i, result)
		}
	}
}

func TestProxyBlocksAfterRateLimit(t *testing.T) {
	app := &Application{}
	p := NewProxy(app, 2)

	// First two requests should succeed
	for i := 1; i <= 2; i++ {
		_, err := p.HandleRequest("http://example.com")
		if err != nil {
			t.Fatalf("request %d should succeed, got error: %v", i, err)
		}
	}

	// Third request should be rate limited
	_, err := p.HandleRequest("http://example.com")
	if err == nil {
		t.Error("expected rate limit error on third request, got nil")
	}
}

func TestProxyRateLimitIsPerURL(t *testing.T) {
	app := &Application{}
	p := NewProxy(app, 1)

	_, err := p.HandleRequest("http://example.com")
	if err != nil {
		t.Fatalf("first URL should succeed: %v", err)
	}

	// Different URL should have its own limit
	_, err = p.HandleRequest("http://other.com")
	if err != nil {
		t.Fatalf("second URL should succeed: %v", err)
	}

	// First URL should now be rate limited
	_, err = p.HandleRequest("http://example.com")
	if err == nil {
		t.Error("expected rate limit error for first URL")
	}
}

func TestProxyImplementsServer(t *testing.T) {
	app := &Application{}
	var s Server = NewProxy(app, 5)
	_, err := s.HandleRequest("http://example.com")
	if err != nil {
		t.Fatalf("proxy as Server should work: %v", err)
	}
}

File: Projects/outOfStockNotification/Design Patterns/Structural/Proxy/ProxyTest.java
package proxy;

import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

class ProxyTest {

    @Test
    void testApplicationHandleRequest() throws RateLimitException {
        Application app = new Application();
        assertEquals("http://example.com - Success", app.handleRequest("http://example.com"));
    }

    @Test
    void testProxyAllowsRequestsWithinLimit() throws RateLimitException {
        Proxy proxy = new Proxy(new Application(), 3);

        for (int i = 1; i <= 3; i++) {
            String result = proxy.handleRequest("http://example.com");
            assertEquals("http://example.com - Success", result);
        }
    }

    @Test
    void testProxyBlocksAfterRateLimit() throws RateLimitException {
        Proxy proxy = new Proxy(new Application(), 2);

        proxy.handleRequest("http://example.com");
        proxy.handleRequest("http://example.com");

        assertThrows(RateLimitException.class, () -> {
            proxy.handleRequest("http://example.com");
        });
    }

    @Test
    void testProxyRateLimitIsPerURL() throws RateLimitException {
        Proxy proxy = new Proxy(new Application(), 1);

        proxy.handleRequest("http://example.com");
        proxy.handleRequest("http://other.com");

        assertThrows(RateLimitException.class, () -> {
            proxy.handleRequest("http://example.com");
        });
    }

    @Test
    void testProxyImplementsServer() throws RateLimitException {
        Server server = new Proxy(new Application(), 5);
        String result = server.handleRequest("http://example.com");
        assertEquals("http://example.com - Success", result);
    }
}
